File: go/zbc/keys.go
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

package zbc

import (
	"crypto/ed25519"
	"crypto/hmac"
	"crypto/pbkdf2"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// ZooBCCoinType is the SLIP-44 coin type of the m/44'/883'/i' wallet path.
const ZooBCCoinType = 883

// KeyPair is a 32-byte seed with its public key.
type KeyPair struct {
	Seed      []byte
	PublicKey []byte
}

// KeyPairFromSeed derives the key pair of a 32-byte seed.
func KeyPairFromSeed(seed []byte) (KeyPair, error) {
	if len(seed) != 32 {
		return KeyPair{}, errors.New("Private key must be 64 hex characters (32 bytes)")
	}
	priv := ed25519.NewKeyFromSeed(seed)
	return KeyPair{Seed: append([]byte{}, seed...), PublicKey: []byte(priv.Public().(ed25519.PublicKey))}, nil
}

// KeyPairFromHex derives the key pair of a 64-hex seed.
func KeyPairFromHex(seedHex string) (KeyPair, error) {
	if !IsHex(seedHex, 64) {
		return KeyPair{}, errors.New("Private key must be 64 hex characters (32 bytes)")
	}
	seed, _ := hex.DecodeString(seedHex)
	return KeyPairFromSeed(seed)
}

// Address is the ZBC_ form of the public key.
func (k KeyPair) Address() string { return MustEncodeZbcAddress(k.PublicKey, "ZBC") }

// NodeAddress is the ZNK_ form of the same key.
func (k KeyPair) NodeAddress() string { return MustEncodeZbcAddress(k.PublicKey, "ZNK") }

// AccountBytes is the 36-byte typed account address: 00000000 || public key.
func (k KeyPair) AccountBytes() []byte { return TypedAddress(TypeZooBC, k.PublicKey) }

// Sign is a detached Ed25519 signature with the seed.
func (k KeyPair) Sign(msg []byte) []byte { return ed25519.Sign(ed25519.NewKeyFromSeed(k.Seed), msg) }

// RandomSeed is a fresh 32-byte seed from crypto/rand.
func RandomSeed() []byte {
	s := make([]byte, 32)
	if _, err := rand.Read(s); err != nil {
		panic(err)
	}
	return s
}

// ValidateMnemonic is true for 12, 15, 18, 21 or 24 English words with a valid checksum.
func ValidateMnemonic(mnemonic string) bool {
	words := strings.Fields(mnemonic)
	switch len(words) {
	case 12, 15, 18, 21, 24:
	default:
		return false
	}
	var bits strings.Builder
	for _, w := range words {
		idx := -1
		for i, x := range BIP39Words {
			if x == w {
				idx = i
				break
			}
		}
		if idx < 0 {
			return false
		}
		bits.WriteString(fmt.Sprintf("%011b", idx))
	}
	s := bits.String()
	cs := len(words) / 3
	entropy := make([]byte, (len(s)-cs)/8)
	for i := range entropy {
		var v byte
		for j := 0; j < 8; j++ {
			v = v<<1 | (s[8*i+j] - '0')
		}
		entropy[i] = v
	}
	sum := sha256.Sum256(entropy)
	return s[len(s)-cs:] == fmt.Sprintf("%08b", sum[0])[:cs]
}

// MnemonicFromEntropy builds a mnemonic from 16–32 bytes of entropy (a multiple of 4).
func MnemonicFromEntropy(entropy []byte) (string, error) {
	if len(entropy) < 16 || len(entropy) > 32 || len(entropy)%4 != 0 {
		return "", errors.New("entropy must be 16-32 bytes, a multiple of 4")
	}
	var bits strings.Builder
	for _, b := range entropy {
		bits.WriteString(fmt.Sprintf("%08b", b))
	}
	sum := sha256.Sum256(entropy)
	bits.WriteString(fmt.Sprintf("%08b", sum[0])[:len(entropy)/4])
	s := bits.String()
	words := make([]string, 0, len(s)/11)
	for i := 0; i+11 <= len(s); i += 11 {
		idx := 0
		for j := 0; j < 11; j++ {
			idx = idx<<1 | int(s[i+j]-'0')
		}
		words = append(words, BIP39Words[idx])
	}
	return strings.Join(words, " "), nil
}

// GenerateMnemonic makes a random mnemonic of 12, 15, 18, 21 or 24 words.
func GenerateMnemonic(words int) (string, error) {
	entropy := make([]byte, (words*11-words/3)/8)
	if _, err := rand.Read(entropy); err != nil {
		return "", err
	}
	return MnemonicFromEntropy(entropy)
}

// MnemonicToSeed is PBKDF2-HMAC-SHA512(mnemonic, "mnemonic"+passphrase, 2048, 64).
func MnemonicToSeed(mnemonic, passphrase string) []byte {
	seed, err := pbkdf2.Key(sha512.New, strings.Join(strings.Fields(mnemonic), " "), []byte("mnemonic"+passphrase), 2048, 64)
	if err != nil {
		panic(err)
	}
	return seed
}

var pathRe = regexp.MustCompile(`^m(/[0-9]+')+$`)

// SLIP10Derive follows a hardened-only Ed25519 path such as m/44'/883'/0' and returns the 32-byte key.
func SLIP10Derive(path string, seed []byte) ([]byte, error) {
	if !pathRe.MatchString(path) {
		return nil, errors.New("invalid derivation path: " + path)
	}
	mac := hmac.New(sha512.New, []byte("ed25519 seed"))
	mac.Write(seed)
	digest := mac.Sum(nil)
	key, chain := digest[:32], digest[32:]
	for _, seg := range strings.Split(path, "/")[1:] {
		var index uint64
		if _, err := fmt.Sscanf(seg[:len(seg)-1], "%d", &index); err != nil || index >= 0x80000000 {
			return nil, errors.New("path index too large")
		}
		data := make([]byte, 37)
		copy(data[1:], key)
		binary.BigEndian.PutUint32(data[33:], uint32(index)+0x80000000)
		mac = hmac.New(sha512.New, chain)
		mac.Write(data)
		digest = mac.Sum(nil)
		key, chain = digest[:32], digest[32:]
	}
	return key, nil
}

// WalletAccount is account index of a mnemonic wallet: m/44'/883'/index'.
func WalletAccount(mnemonic string, index int, passphrase string) (KeyPair, string, error) {
	path := fmt.Sprintf("m/44'/%d'/%d'", ZooBCCoinType, index)
	key, err := SLIP10Derive(path, MnemonicToSeed(mnemonic, passphrase))
	if err != nil {
		return KeyPair{}, "", err
	}
	kp, err := KeyPairFromSeed(key)
	return kp, path, err
}

File: ts/src/keys.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
/** Keys: seed -> key pair -> addresses; BIP-39 + SLIP-10 wallets (spec/signing.md section 1). */
import { publicKeyFromSeed } from "./crypto/ed25519.js";
import { hmacSha512, pbkdf2Sha512, sha256 } from "./crypto/sha2.js";
import { encodeZbcAddress, typedAddress, AccountType } from "./address.js";
import { bytesToHex, concat, hexToBytes, isHex, utf8 } from "./util/bytes.js";
import { BIP39_WORDS } from "./util/bip39-words.js";

export interface KeyPair {
  /** The 32-byte seed (what the tools call the private key). */
  seed: Uint8Array;
  publicKey: Uint8Array;
  /** `ZBC_` form of the public key. */
  address: string;
  /** `ZNK_` form, the node-key spelling of the same key. */
  nodeAddress: string;
  /** 36-byte typed account address: 00000000 ‖ public key. */
  accountBytes: Uint8Array;
}

export function keyPairFromSeed(seed: Uint8Array | string): KeyPair {
  const s = typeof seed === "string" ? seedFromHex(seed) : seed;
  if (s.length !== 32) throw new Error("Private key must be 64 hex characters (32 bytes)");
  const publicKey = publicKeyFromSeed(s);
  return { seed: s, publicKey, address: encodeZbcAddress(publicKey, "ZBC"), nodeAddress: encodeZbcAddress(publicKey, "ZNK"),
           accountBytes: typedAddress(AccountType.ZooBC, publicKey) };
}

export function seedFromHex(hex: string): Uint8Array {
  if (!isHex(hex, 64)) throw new Error("Private key must be 64 hex characters (32 bytes)");
  return hexToBytes(hex);
}

/** A fresh random seed from the platform's CSPRNG (Node and browsers). */
export function randomSeed(): Uint8Array {
  const s = new Uint8Array(32);
  globalThis.crypto.getRandomValues(s);
  return s;
}

// ---- BIP-39 and SLIP-10 -----------------------------------------------------------------------
/** True for 12, 15, 18, 21 or 24 words of the English list with a valid checksum. */
export function validateMnemonic(mnemonic: string): boolean {
  const words = mnemonic.trim().split(/\s+/);
  if (![12, 15, 18, 21, 24].includes(words.length)) return false;
  let bits = "";
  for (const w of words) {
    const i = BIP39_WORDS.indexOf(w);
    if (i < 0) return false;
    bits += i.toString(2).padStart(11, "0");
  }
  const cs = words.length / 3;
  const entropy = new Uint8Array((bits.length - cs) / 8);
  for (let i = 0; i < entropy.length; i++) entropy[i] = parseInt(bits.slice(8 * i, 8 * i + 8), 2);
  const hash = sha256(entropy);
  return bits.slice(bits.length - cs) === hash[0].toString(2).padStart(8, "0").slice(0, cs);
}

/** Entropy (16–32 bytes, multiple of 4) -> mnemonic. */
export function mnemonicFromEntropy(entropy: Uint8Array): string {
  if (entropy.length < 16 || entropy.length > 32 || entropy.length % 4 !== 0) throw new Error("entropy must be 16-32 bytes");
  let bits = "";
  for (const b of entropy) bits += b.toString(2).padStart(8, "0");
  const cs = entropy.length / 4;
  bits += sha256(entropy)[0].toString(2).padStart(8, "0").slice(0, cs);
  const words: string[] = [];
  for (let i = 0; i < bits.length / 11; i++) words.push(BIP39_WORDS[parseInt(bits.slice(11 * i, 11 * i + 11), 2)]);
  return words.join(" ");
}

export function generateMnemonic(words: 12 | 15 | 18 | 21 | 24 = 24): string {
  const entropy = new Uint8Array((words * 11 - words / 3) / 8);
  globalThis.crypto.getRandomValues(entropy);
  return mnemonicFromEntropy(entropy);
}

/** BIP-39 seed: PBKDF2-HMAC-SHA512(mnemonic, "mnemonic" + passphrase, 2048, 64). */
export function mnemonicToSeed(mnemonic: string, passphrase = ""): Uint8Array {
  return pbkdf2Sha512(utf8(mnemonic.trim().split(/\s+/).join(" ")), utf8("mnemonic" + passphrase), 2048, 64);
}

/** SLIP-10 for Ed25519 along a hardened-only path such as m/44'/883'/0'. Returns the 32-byte key. */
export function slip10Derive(path: string, seed: Uint8Array): Uint8Array {
  if (!/^m(\/[0-9]+')+$/.test(path)) throw new Error("invalid derivation path: " + path);
  let I = hmacSha512(utf8("ed25519 seed"), seed);
  let key = I.slice(0, 32), chain = I.slice(32);
  for (const seg of path.split("/").slice(1)) {
    const index = Number(seg.slice(0, -1));
    if (index >= 0x80000000) throw new Error("path index too large");
    const data = new Uint8Array(37);
    data.set(key, 1);
    new DataView(data.buffer).setUint32(33, index + 0x80000000, false);
    I = hmacSha512(chain, data);
    key = I.slice(0, 32); chain = I.slice(32);
  }
  return key;
}

export const ZOOBC_COIN_TYPE = 883;
export interface WalletAccount extends KeyPair { index: number; path: string }

/** Account `index` of a mnemonic wallet: m/44'/883'/index'. */
export function walletAccount(mnemonic: string, index: number, passphrase = ""): WalletAccount {
  const path = `m/44'/${ZOOBC_COIN_TYPE}'/${index}'`;
  const seed = slip10Derive(path, mnemonicToSeed(mnemonic, passphrase));
  return { ...keyPairFromSeed(seed), index, path };
}

export const seedHex = (kp: KeyPair): string => bytesToHex(kp.seed);
export const publicKeyHex = (kp: KeyPair): string => bytesToHex(kp.publicKey);
export { concat };
